modules/general_functions.py

```diff
@@ -656,3 +657,51 @@ def find_next(iterable: Iterable[T], cond: Callable[[T], bool]) -> T:
         return next(i for i in iterable if cond(i))
     except StopIteration:
         raise ValueError("Value not found in iterable.")
+
+# FIXME: Find better placement
+class ReferenceDensity:
+    def __init__(self, layers: List = None):
+        self.layers = layers
+        self.reference_density = 0.0
+        self.layer_reference_density = 0.0
+        self.reference_density_limit = 10.0
+        self.reference_density_meter = 10.0
+        self.masses: List = []
+        self.g_mass = 0.0
+
+    def update_reference_density(self):
+        for layer in self.layers:
+            self.masses = []
+            # If there is only one layer
+            if len(self.layers) == 1:
+                self.calculate_reference_density(layer)
+
+            # If the layer's width is smaller than 10 nm
+            elif self.reference_density_limit - layer.thickness >= 0:
+                self.reference_density_meter = self.reference_density_meter - \
+                                          layer.thickness
+                self.calculate_reference_density(layer)
+
+            # If the layer is the last one before 10nm limit
+            elif layer.thickness > self.reference_density_limit:
+                self.calculate_reference_density(layer)
+                break
+
+    def calculate_reference_density(self, layer):
+        layer_densities = []
+        for element in layer.elements:
+            self.g_mass = 0.0
+            mass = element.get_mass()
+            self.masses.append(mass * element.amount)
+            for mass in self.masses:
+                self.g_mass = (convert_amu_to_kg(mass) * 1000)
```

Comment:
There seems to be a missing `+=` operator.

I don't understand this inner loop. Why are the previous elements' masses iterated multiple times? The first elements' mass is summed `n` times, the next one's `n-1`, ..., and the last one's `1` time. Shouldn't they all be summed exactly once?

Reply:
OK